File: SRC/JAVA/ProcessingGPSData-Java/src/SentenceHandler.java
/**
 * SentenceHandler deals with taking in sentences from
 * the StreamReaders, determines what type of sentences
 * they are and then assigns data to the correct place
 * in order to be used in plotting a route.
 *
 * @author - James Euesden <jee22@aber.ac.uk>
 * @version - 1.0
 */
public class SentenceHandler {

    /**
     * Start the class with a new list for the satellites.
     */
    public SentenceHandler(){

    }

    /**
     * Sent a sentence to the DECONSTRUCTION.
     * @param currentSentence The sentence to be parsed.
     */
    protected void parse(Stream streamer, String currentSentence){
        deconstructSentence(streamer, currentSentence);
    }

    /**
     * Takes a sentence and finds out what type it is and
     * passes it to the appropriate method.
     * @param sentence The sentence to be deconstructed.
     */
    private void deconstructSentence(Stream streamer, String sentence){
        if(sentence != null){
            String[] sentenceComponents = sentence.split(",");
            if (sentenceComponents[0].equals("$GPRMC")){
                handleGPRMC(streamer, sentenceComponents);
            } else if (sentenceComponents[0].equals("$GPGGA")){
                handleGPGGA(streamer, sentenceComponents);
            } else if (sentenceComponents[0].equals("$GPGSV")){
                handleGPGSV(streamer, sentenceComponents);
            }
        }
    }

    /**
     *
     * @param sentenceComponents The full sentence.
     */
    private void handleGPRMC(Stream streamer, String[] sentenceComponents){
        /**
         * Set the date and time. The GPRMC is the only sentence, aside GPZDA,
         * that provides a Date for the timestamps..
         */
        streamer.setDateAndTime(sentenceComponents[1], sentenceComponents[9]);
        streamer.setLatitude(sentenceComponents[3], sentenceComponents[4].charAt(0));
        streamer.setLongitude(sentenceComponents[5], sentenceComponents[6].charAt(0));
    }

    /**
     *
     * @param sentenceComponents The full sentence.
     */
    private void handleGPGGA(Stream streamer, String[] sentenceComponents){
        streamer.updateTime(sentenceComponents[1]);
        streamer.setLatitude(sentenceComponents[2], sentenceComponents[3].charAt(0));
        streamer.setLongitude(sentenceComponents[4], sentenceComponents[5].charAt(0));
        streamer.setElevation(sentenceComponents[9]);
    }

    /**
     *
     * @param sentenceComponents The full sentence.
     */
    private void handleGPGSV(Stream streamer, String[] sentenceComponents){
        /**
         * Looks to see if this is the first sentence of the current group. If it
         * is, then the amounts of SNR are reset.
         */
        if(sentenceComponents[2].equals("1")){
            streamer.set35AndAboveSNR(0);
            streamer.setSignalsBetween30And35(0);
        }
        String snr;
        /**
         * For all of the amounts of Satellites and SNR in the sentence,
         * go through and send them to be added into the Stream. First
         * check that they are not the last of the line, by looking
         * for the asterisk of the check sum.
         *
         * If the asterisk is there, get everything from before it.
         */
        for(int i = 7; i < sentenceComponents.length; i+=4){
            snr = sentenceComponents[i];
            if(snr.contains("*")){
                int asterisk = snr.indexOf('*');
                snr = snr.substring(0, asterisk);
            }

            if(!snr.equals("")){
                streamer.calcSNRAmounts(snr);
            }
        }
    }
}

File: SRC/Cplusplus/GPSProcessing_Cplusplus/SentenceHandler.cpp
/* 
 * File:   SentenceHandler.cpp
 * Author: jee22
 *
 * Takes in sentences and determines what
 * they are and what to do with them.
 */

#include <string>
#include <iostream>
#include <sstream>
#include <time.h>
#include "SentenceHandler.h"
#include "Stream.h"

using namespace std;

SentenceHandler::SentenceHandler() {
}

SentenceHandler::SentenceHandler(const SentenceHandler& orig) {
}

SentenceHandler::~SentenceHandler() {
}

/**
 * Takes a GPRMC sentence and takes the needed data.
 * @param streamer The stream for the sentence.
 * @param sentence The sentence itself.
 */
void SentenceHandler::handle_gprmc(Stream * streamer, stringstream * sentence){
    string dont_care;
    string time_string;
    string date_string;
    string location; /* For Latitude, Longitude and Elevation.*/
    string facing; /* N,E,S,W */
    
    getline(*sentence, time_string, ','); /* Get time */
    
    getline(*sentence, dont_care, ','); /* Ignore */
    
    getline(*sentence, location, ','); /* Get Latitude */
    getline(*sentence, facing, ','); /* Get facing */
    streamer->setLatitude(location, facing);
    
    getline(*sentence, location, ','); /* Get Longitude */
    getline(*sentence, facing, ','); /* Get facing */
    streamer->setLongitude(location, facing);

    getline(*sentence, dont_care, ','); /* Ignore */
    getline(*sentence, dont_care, ','); /* Ignore */
    
    getline(*sentence, date_string, ','); /* Get date */
    streamer->setDateAndTime(date_string, time_string);
}

/**
 * Takes a GPGGA sentence and gets the needed data from it.
 * @param streamer The stream where the sentence came from.
 * @param sentence The sentence itself.
 */
void SentenceHandler::handle_gpgga(Stream * streamer, stringstream * sentence){
    string dont_care;
    string time_string;
    string location;
    string facing;
    
    getline(*sentence, time_string, ','); /* Get time */
    streamer->updateTime(time_string);
    
    getline(*sentence, location, ','); /* Get Latitude */
    getline(*sentence, facing, ','); /* Get facing (N/S) */
    streamer->setLatitude(location, facing);
    
    getline(*sentence, location, ','); /* Get Longitude */
    getline(*sentence, facing, ','); /* Get facing (E/W) */
    streamer->setLongitude(location, facing);
    
    getline(*sentence, dont_care, ','); /* Ignore */
    getline(*sentence, dont_care, ','); /* Ignore */
    getline(*sentence, dont_care, ','); /* Ignore */
    
    getline(*sentence, location, ','); /* Elevation */
    streamer->setElevation(location);

    streamer->updateTime(time_string);    
}

/**
 * Dealing with a GPGSV, and getting the data needed.
 * @param streamer The stream the sentence came from.
 * @param sentence The sentence itself.
 */
void SentenceHandler::handle_gpgsv(Stream * streamer, stringstream * sentence){
    string dont_care;
    getline(*sentence, dont_care, ','); /* Ignore - Total number of GSVs for this set. */
    getline(*sentence, dont_care, ','); /* Sentence Number */
    int sentence_number;
    stringstream ss(dont_care);
    ss >> sentence_number;
    getline(*sentence, dont_care, ','); /* Ignore - Total number of satellites */

    /* If the sentence is the first, start the stream recordings
     for the satellites from 0 again. */
    if (sentence_number == 1) {
        streamer->set35AndAboveSNR(0);
        streamer->setBetween30And35SNR(0);
    }

    int last_line = 0;
    /* While the last line hasn't been reached yet, keep going over the
     sentence. */
    while (last_line != 1) {
        getline(*sentence, dont_care, ','); /* Ignore - Satellite PRN */
        getline(*sentence, dont_care, ','); /* Ignore - Elevation */
        getline(*sentence, dont_care, ','); /* Ignore - Azimuth */
        
        string snr;
        getline(*sentence, snr, ','); /* SNR value */
        
        int asterisk = snr.find('*');
        
        if (asterisk != string::npos) {
            /* Get the value before the asterisk and check
             to show that this is the last section of the line. */
            snr = snr.substr(0, asterisk);
            last_line = 1;
        }

        /* If the snr value is higher than 0, increment the
         number of satellites through a call to calc_snr_amounts()*/
        if (snr.compare("") != 0) {
            streamer->calcSNRAmounts(snr);
        }
    }
}

void SentenceHandler::parseSentence(Stream * streamer, string sentence){
    stringstream linestream(sentence);
    string sentence_signifier;
    getline(linestream, sentence_signifier, ',');
    if(sentence_signifier.compare("$GPRMC") == 0){
        handle_gprmc(streamer, &linestream);
    } else if(sentence_signifier.compare("$GPGGA") == 0){
        handle_gpgga(streamer, &linestream);
    } else if(sentence_signifier.compare("$GPGSV") == 0){
        handle_gpgsv(streamer, &linestream);
    }
}
